File: androidApp/src/main/java/com/pm/lovecalendar_kmm/android/presentation/MeetingsScreen.kt
package com.pm.lovecalendar_kmm.android.presentation

import android.widget.Toast
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.toUpperCase
import androidx.compose.ui.unit.dp
import com.pm.lovecalendar_kmm.android.presentation.components.DaysLeftItem
import com.pm.lovecalendar_kmm.android.presentation.components.HelperButton
import com.pm.lovecalendar_kmm.android.presentation.components.MonthItem
import com.pm.lovecalendar_kmm.presentation.MeetingsEvent
import com.pm.lovecalendar_kmm.presentation.MeetingsState
import com.pm.lovecalendar_kmm.android.R

@Composable
fun MeetingsScreen(
    state: MeetingsState,
    onEvent: (MeetingsEvent) -> Unit
) {
    val context = LocalContext.current
    val toastMessage = stringResource(R.string.feature_coming_soon)

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(MaterialTheme.colors.background)
            .padding(horizontal = 16.dp)
            .verticalScroll(rememberScrollState())
    ) {
        Row(
            modifier = Modifier.padding(top = 16.dp, bottom = 20.dp)
        ) {
            HelperButton(
                onClick = { Toast.makeText(context, toastMessage, Toast.LENGTH_LONG).show() },
                drawablePath = R.drawable.settings
            )
            Spacer(modifier = Modifier.weight(1f))
            DaysLeftItem(text = if (state.isInEditMode) stringResource(R.string.edit).uppercase() else state.daysLeftText)
            Spacer(modifier = Modifier.weight(1f))
            HelperButton(
                onClick = { onEvent(MeetingsEvent.ToggleEditingMode) },
                drawablePath = if (state.isInEditMode) R.drawable.done else R.drawable.add
            )
        }
        MonthItem(
            isEditing = state.isInEditMode,
            isCurrent = true,
            month = state.firstMonth.name,
            year = state.firstYear.toString(),
            firstDayPosition = state.firstMonthFirstDayOfWeekPosition,
            emptyDatesAmount = state.firstMonthEmptyDatesAmount,
            dates = state.firstMonthDates,
            onDateTap = { onEvent(MeetingsEvent.OnDateTap(it)) }
        )
        Spacer(modifier = Modifier.height(16.dp))
        MonthItem(
            isEditing = state.isInEditMode,
            isCurrent = false,
            month = state.secondMonth.name,
            year = state.secondYear.toString(),
            firstDayPosition = state.secondMonthFirstDayOfWeekPosition,
            emptyDatesAmount = state.secondMonthEmptyDatesAmount,
            dates = state.secondMonthDates,
            onDateTap = { onEvent(MeetingsEvent.OnDateTap(it)) }
        )
        Spacer(modifier = Modifier.height(16.dp))
    }
}

File: iosApp/iosApp/presentation/MeetingsScreen.swift
//
//  MeetingsScreen.swift
//  iosApp
//
//  Created by Paul on 27/01/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import SwiftUI
import shared

struct MeetingsScreen: View {
    private var meetingsDataSource: MeetingsDataSource
    private var generateDates: GenerateDates
    private var findFirstDayOfMonthPosition: FindFirstDayOfMonthPosition
    @ObservedObject var viewModel: IOSMeetingsViewModel
    
    init(
        meetingsDataSource: MeetingsDataSource,
        generateDates: GenerateDates,
        findFirstDayOfMonthPosition: FindFirstDayOfMonthPosition
    ) {
        self.meetingsDataSource = meetingsDataSource
        self.generateDates = generateDates
        self.findFirstDayOfMonthPosition = findFirstDayOfMonthPosition
        self.viewModel = IOSMeetingsViewModel(
            meetingsDataSource: meetingsDataSource,
            generateDates: generateDates,
            findFirstDayOfMonthPosition: findFirstDayOfMonthPosition
        )
    }
    
    var body: some View {
        let screenRect = UIScreen.main.bounds
        let screenWidth = screenRect.size.width
        let padding = (screenWidth - 360 - 32) / 2
        
        VStack {
            HStack {
                HelperButton(
                    onClick: {},
                    imageName: "gearshape"
                )
                Spacer()
                DateItem(text: viewModel.state.daysLeftText)
                Spacer()
                HelperButton(
                    onClick: { viewModel.onEvent(event: MeetingsEvent.ToggleEditingMode()) },
                    imageName: viewModel.state.isInEditMode ? "checkmark" : "plus"
                )
            }
            .padding(.top, 16)
            .padding(.bottom, 20)
            .padding(.leading, padding)
            .padding(.trailing, padding)
            MonthItem(
                isEditing: viewModel.state.isInEditMode,
                isCurrent: true,
                month: viewModel.state.firstMonth.name,
                year: String(viewModel.state.firstYear),
                firstDayPosition: Int(viewModel.state.firstMonthFirstDayOfWeekPosition),
                emptyDatesAmount: Int(viewModel.state.firstMonthEmptyDatesAmount),
                dates: viewModel.state.firstMonthDates,
                onDateTap: { date in viewModel.onEvent(event: MeetingsEvent.OnDateTap(appDate: date)) }
            )
            .padding(.bottom, 16)
            MonthItem(
                isEditing: viewModel.state.isInEditMode,
                isCurrent: false,
                month: viewModel.state.secondMonth.name,
                year: String(viewModel.state.secondYear),
                firstDayPosition: Int(viewModel.state.secondMonthFirstDayOfWeekPosition),
                emptyDatesAmount: Int(viewModel.state.secondMonthEmptyDatesAmount),
                dates: viewModel.state.secondMonthDates,
                onDateTap: { date in viewModel.onEvent(event: MeetingsEvent.OnDateTap(appDate: date)) }
            )
            Spacer()
        }
        .padding(.top, 1)
        .padding(.trailing, 16)
        .padding(.leading, 16)
        .onAppear {
            viewModel.startObserving()
        }
        .onDisappear {
            viewModel.dispose()
        }
    }
}

